modules/core/include/opencv2/core/hal/intrin_avx512.hpp

```diff
@@ -0,0 +1,2674 @@
+
+OPENCV_HAL_IMPL_AVX512_REDUCE_32(min, min, v_uint16x32, ushort, epu16)
+OPENCV_HAL_IMPL_AVX512_REDUCE_32(min, min, v_int16x32,  short,  epi16)
+OPENCV_HAL_IMPL_AVX512_REDUCE_32(max, max, v_uint16x32, ushort, epu16)
+OPENCV_HAL_IMPL_AVX512_REDUCE_32(max, max, v_int16x32,  short,  epi16)
+OPENCV_HAL_IMPL_AVX512_REDUCE_32_SUM(v_uint16x32, uint, epu16)
+OPENCV_HAL_IMPL_AVX512_REDUCE_32_SUM(v_int16x32,  int,  epi16)
+
+#define OPENCV_HAL_IMPL_AVX512_REDUCE_64(func, ifunc, _Tpvec, sctype, suffix)                                      \
+    inline sctype v_reduce_##func(const _Tpvec& a)                                                                 \
+    {                                                                                                              \
+        __m256i half = _mm256_##ifunc##_##suffix(_v512_extract_low(a.val), _v512_extract_high(a.val));             \
+        __m128i quarter = _mm_##ifunc##_##suffix(_mm256_castsi256_si128(half), _mm256_extracti128_si256(half, 1)); \
+        return (sctype)_mm512_reduce_##ifunc##_epi32(_mm512_cvt##suffix##_epi32(quarter));                         \
+    }
+#define OPENCV_HAL_IMPL_AVX512_REDUCE_64_SUM(_Tpvec, sctype, suffix)                            \
+    inline sctype v_reduce_sum(const _Tpvec& a)                                                 \
+    {                                                                                           \
+        __m512i half = _mm512_add_epi16(_mm512_cvt##suffix##_epi16(_v512_extract_low(a.val)),   \
+                                        _mm512_cvt##suffix##_epi16(_v512_extract_high(a.val))); \
+        __m256i quarter = _mm256_add_epi16(_v512_extract_low(half), _v512_extract_high(half));  \
+        return (sctype)_mm512_reduce_add_epi32(_mm512_cvtepi16_epi32(quarter));                 \
+    }
+
+OPENCV_HAL_IMPL_AVX512_REDUCE_64(min, min, v_uint8x64, uchar, epu8)
+OPENCV_HAL_IMPL_AVX512_REDUCE_64(min, min, v_int8x64,  char,  epi8)
+OPENCV_HAL_IMPL_AVX512_REDUCE_64(max, max, v_uint8x64, uchar, epu8)
+OPENCV_HAL_IMPL_AVX512_REDUCE_64(max, max, v_int8x64,  char,  epi8)
+OPENCV_HAL_IMPL_AVX512_REDUCE_64_SUM(v_uint8x64, uint, epu8)
+OPENCV_HAL_IMPL_AVX512_REDUCE_64_SUM(v_int8x64,  int,  epi8)
+
+inline v_float32x16 v_reduce_sum4(const v_float32x16& a, const v_float32x16& b,
+                                  const v_float32x16& c, const v_float32x16& d)
+{
+    __m256 abl = _mm256_hadd_ps(_v512_extract_low(a.val), _v512_extract_low(b.val));
+    __m256 abh = _mm256_hadd_ps(_v512_extract_high(a.val), _v512_extract_high(b.val));
+    __m256 cdl = _mm256_hadd_ps(_v512_extract_low(c.val), _v512_extract_low(d.val));
+    __m256 cdh = _mm256_hadd_ps(_v512_extract_high(c.val), _v512_extract_high(d.val));
+    return v_float32x16(_v512_combine(_mm256_hadd_ps(abl, cdl), _mm256_hadd_ps(abh, cdh)));
+}
+
+inline unsigned v_reduce_sad(const v_uint8
```

Comment:
Abstraction layer should be efficient, so lets use **native** mask instead:
- `__mmask64`
- `__mmask32`
- `__mmask16`
- `__mmask8`

See https://github.com/opencv/opencv/wiki/OE-29.-Adding-AVX512-Support for proposal details

Reply:
Additionally suggest removing lane types from names (from proposal):
- `v_mask_uint8x32` => `v_mask32`
- `v_mask_float32x16` => `v_mask16`

Reply:
BTW, GitHub has been bugged again, related code is about `v_any()`/`v_all()` functions.

Reply:
v_check_any() and v_check_all() reworked. I think it doesn't make sense to add support of masks to the interface of these functions since the result will anyway be casted to bool immediately. IMO support of v_mask_smth should be added to interface of comparison operators and v_select. I'm going to address it as a separate PR.